### PR TITLE
Add DynamicIndexer for dynamic top level columns

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: server/src/main/java/io/crate/execution/dml/DynamicIndexer.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+     * </p>
+     * <p>
+     *  Users should sanitize the value using {@link DataType#sanitizeValue(Object)}
+     *  of the result type.
+     * </p>
+     */
+    static DataType<?> guessType(Object value) {
+        DataType<?> type = DataTypes.guessType(value);
+        return switch (type.id()) {
+            case ShortType.ID, IntegerType.ID -> DataTypes.LONG;
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/dml/DynamicIndexer.java:129]: `Byte` also should become `Long`?

File: server/src/main/java/io/crate/execution/dml/DynamicIndexer.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            if (storageSupport == null) {
+                throw new IllegalArgumentException(
+                    "Cannot create columns of type " + type.getName() + " dynamically. " +
+                    "Storage is not supported for this type");
+            }
+            boolean nullable = true;
+            int position = -1;
+            Symbol defaultExpression = null;
+            Reference newColumn = new SimpleReference(
+                refIdent,
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/dml/DynamicIndexer.java:86]: `DynamicReference` probably fits better for `DynamicIndexer`?

Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/dml/DynamicIndexer.java:86]: Why? At this point it stops being dynamic

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/DynamicIndexer.java:86]: OK, I misunderstood the point of `DynamicReference` - I thought we always distinguish that some column was created dynamically. Apparently it's just intermediate thing.

File: server/src/main/java/io/crate/execution/dml/DynamicIndexer.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/dml/DynamicIndexer.java]: Done

**Review by BaurzhanSakhariev — APPROVED** [review]

**Pull request merged**